### DISCUSSION TITLE
ERR_TOO_MANY_REDIRECTS in Chrome

### DISCUSSION THREAD

**Discussion by profcaju** [YOURLS/YOURLS#3402] [Q&A]
Three days ago I received an user message stating that he could not open any link from my yourls website, because his chrome was giving this ERR_TOO_MANY_REDIRECTS.

Testing in my machine, with Firefox, I could not replicate this problem. Same with Safari (I'm using MacOs).

But, when trying with Google Chrome, I received the same error! 

![image](https://user-images.githubusercontent.com/24816099/183089838-d99ba097-809f-4a84-9e82-96b19273bf56.png)

Even clearing cookies, cache, everything, the problem persists.

I think that this might be a local problem, because when I change my DNS Server (from 8.8.8.8 to my local ISP DNS server), I can access my URL in Chrome with no problem. And if I change my DNS server to Cloudflare's 1.1.1.1, I get the same problem.

This problem occurs only in Chrome. Using Safari or Firefox I can use any DNS Server that my yourls installation will work.

I already uninstall Chrome and installed again with no modification in this behavior.

How could I start an investigation about this problem?

Kind regards,

**Comment by ozh** [comment]
What does server logs say when requesting a short URL that trigger the ERR_TOO_MANY ?

**Reply by profcaju** [reply]
Hello, Ozh.

There is nothing being logged in the Apache error log. Looks like the request never reached the server.

Thinking about this behavior, I realized that the culprit could be Cloudflare.

So, I started looking around Cloudflare, and when enabling SSL through Cloudflare (Flexible option), the problem disappear!!!!!! Now I can connect to my yourls installation with any browser and any DNS Server.
Looks like the loop of redirection were between client and Cloudflare, never reaching my web server.

What is strange is the fact that this was happening only with Google Chrome :(

**Comment by logusivam** [comment]
The combination of symptoms you've described — Chrome-only, DNS-server-dependent, persistent after clearing cache — points to a very specific root cause. This is almost certainly an **HSTS (HTTP Strict Transport Security) conflict** interacting with how Chrome handles HTTPS enforcement differently from Firefox and Safari.

---

## Why Chrome behaves differently from Firefox and Safari here

Chrome has its own built-in **HSTS preload list** and handles HSTS headers more aggressively than Firefox or Safari in several ways:

1. Chrome maintains a persistent HSTS cache that **survives a full cache clear** unless you clear it specifically through `chrome://net-internals/#hsts`
2. Chrome enforces HSTS for domains on Google's own preload list, which it ships with — meaning it can enforce HTTPS even on first visit before seeing any HSTS header
3. Chrome uses Google's DNS-over-HTTPS (DoH) by default when you set 8.8.8.8 or 1.1.1.1 as your DNS resolver — your ISP's DNS typically bypasses DoH

That third point is the key to why switching DNS servers changes the behavior: Chrome's DoH (Secure DNS) resolves differently than your ISP's plain DNS, and the resulting IP or connection path triggers a different redirect behavior.

---

## The most likely root cause — mixed HTTP/HTTPS redirect loop

Here's what's almost certainly happening:

```
Chrome (with DoH via 8.8.8.8/1.1.1.1):
1. Browser requests http://yourdomain.com/shortlink
2. HSTS forces upgrade to https://yourdomain.com/shortlink
3. YOURLS/server redirects back to http:// for some reason (config mismatch)
4. HSTS forces upgrade again → infinite loop → ERR_TOO_MANY_REDIRECTS

Chrome (with ISP DNS, no DoH):
1. Browser requests http://yourdomain.com/shortlink
2. DoH is bypassed → HSTS enforcement may differ
3. Loop doesn't trigger
```

The underlying YOURLS/server misconfiguration causes a redirect loop that Chrome's stricter HSTS enforcement catches, while Firefox and Safari let the same misconfiguration pass (they're more tolerant of redirect chains before giving up).

---

## Step 1 — Check Chrome's HSTS cache for your domain

This is the first thing to check and the quickest to rule in or out.

1. Open Chrome and go to: `chrome://net-internals/#hsts`
2. In the **"Query HSTS/PKP domain"** section, type your YOURLS domain (e.g. `yourdomain.com`)
3. Click **Query**

If it returns HSTS data (showing `static_sts_domain`, `dynamic_sts_domain`, or similar), your domain has HSTS entries stored in Chrome. This alone doesn't explain the loop, but confirms Chrome is enforcing HTTPS independently.

To clear it:

1. In the **"Delete domain security policies"** section on the same page
2. Enter your domain and click **Delete**
3. Test immediately — if the error disappears, HSTS cache was part of the chain

---

## Step 2 — Check for the actual redirect chain

Open Chrome DevTools → **Network** tab → enable **"Preserve log"** → visit your short URL. You'll see every redirect hop. Look for:

- Is it bouncing between `http://` and `https://`?
- Is there a redirect to `www.` and back to non-`www.` (or vice versa)?
- Is there a trailing slash redirect loop?

The specific sequence will tell you exactly where the loop originates.

Alternatively, use `curl` to trace the full redirect chain from the command line:

```bash
curl -v -L --max-redirs 10 http://yourdomain.com/yourshortlink 2>&1 | grep -E "^[<>*]"
```

and with HTTPS:

```bash
curl -v -L --max-redirs 10 https://yourdomain.com/yourshortlink 2>&1 | grep -E "^[<>*]"
```

Look for any redirect that sends you from HTTPS back to HTTP — that's the loop trigger.

---

## Step 3 — Check your YOURLS config for HTTP/HTTPS mismatch

Open `user/config.php` and check:

```php
define( 'YOURLS_SITE', 'https://yourdomain.com' );  // must match your actual serving protocol exactly
```

If `YOURLS_SITE` is set to `http://` but your server forces HTTPS (via `.htaccess`, Nginx config, or a load balancer/CDN), YOURLS will generate redirect targets with `http://` while your server immediately upgrades them back to `https://` — creating a redirect loop that Chrome's HSTS enforcement amplifies into `ERR_TOO_MANY_REDIRECTS`.

**This is the single most common cause of this exact symptom.**

Fix: ensure `YOURLS_SITE` uses `https://` and matches your domain exactly (including `www.` or non-`www.` preference):

```php
define( 'YOURLS_SITE', 'https://yourdomain.com' );   // no trailing slash
```

---

## Step 4 — Check for proxy/load balancer header issues

If your YOURLS install sits behind a **reverse proxy, load balancer, CDN (Cloudflare), or Docker networking layer**, the proxy may be terminating SSL and forwarding plain HTTP to YOURLS internally. YOURLS then sees an HTTP request, generates HTTP redirect targets, which the proxy or browser then upgrades to HTTPS again — loop.

Check whether YOURLS is receiving the correct forwarded protocol. In your `.htaccess`, you may need to add:

```apache
# Trust the X-Forwarded-Proto header from your proxy
RewriteCond %{HTTP:X-Forwarded-Proto} =http
RewriteRule ^ https://%{HTTP_HOST}%{REQUEST_URI} [L,R=301]
```

Or in `user/config.php`, force YOURLS to treat the connection as HTTPS when behind a proxy:

```php
// Add this if behind a proxy that terminates SSL
if ( isset( $_SERVER['HTTP_X_FORWARDED_PROTO'] ) && $_SERVER['HTTP_X_FORWARDED_PROTO'] === 'https' ) {
    $_SERVER['HTTPS'] = 'on';
}
```

---

## Why DNS server choice changes the behavior

The reason switching from 8.8.8.8/1.1.1.1 to your ISP DNS "fixes" it in Chrome specifically:

- Chrome automatically enables **DNS-over-HTTPS (Secure DNS)** when you use known DoH-capable resolvers like Google (8.8.8.8) and Cloudflare (1.1.1.1)
- With DoH active, Chrome also enables stricter **connection handling** including more aggressive HSTS enforcement and connection coalescing
- Your ISP's DNS is plain DNS (unencrypted UDP/TCP), which doesn't trigger Chrome's DoH mode — and without DoH, Chrome's HSTS enforcement is slightly less aggressive, which is why the loop doesn't trigger with the ISP DNS

This DNS behavior difference is a **diagnostic signal**, not the root cause. The root cause is the redirect loop itself — Chrome's DoH just makes it visible by enforcing HTTPS more strictly.

---

## Investigation checklist — work through in order

| Step | What to do | What to look for |
|------|-----------|-----------------|
| 1 | `chrome://net-internals/#hsts` → Query your domain | HSTS entries present; try deleting and retesting |
| 2 | Chrome DevTools Network tab with Preserve Log → visit short URL | HTTP↔HTTPS bounce, `www` vs non-`www` loop, trailing slash loop |
| 3 | Check `YOURLS_SITE` in `user/config.php` | Must be `https://` not `http://`, must match actual serving URL exactly |
| 4 | Check for reverse proxy/CDN SSL termination | `X-Forwarded-Proto` header handling |
| 5 | Check `.htaccess` for conflicting HTTPS redirect rules | Multiple rules both redirecting to HTTPS creating a loop |

Start with Step 2 (the curl redirect trace) — seeing the actual redirect chain will immediately tell you where the loop originates and make all the other steps obvious.

Hope this helps — if it answers your question, would you mind clicking **"Mark as answer"** so others searching for the same thing can find it quickly?
